Pa11ycrawler updates for AC-709

michaelyoungstrom opened the pull request:
Purpose of this PR:

(AC-709). Allow pa11ycrawler to crawl a single url, and dump the response through the pipeline.

Commit:
Enable pa11ycrawler to crawl single url

single_url isn't needed inside spider

Overwriting Parse is no longer necessary. Allowing crawl to access one url

fix whitespace issue

more whitespace issues

Add test:

keep spacing uniform

add support for when no port is specified, and allow logging error type

update README

Make sure new feature runs with custom credentials

edit test and add a space for make quality

Fixing make quality errors

support python 2 and 3. And utilize handle_error for all single_url calls

change README

comment urlparse import

Update logging and expand handle_error

fix make quality errors

more make quality errors

Pin Twisted version

fix spacing on log output

testing for handle_error

redundant slash

make quality issues..

style change and remove print

Add testfixtures to dev requirements

Keep bracket in test file

comment mock_check method

keep empty line

benpatterson reviewed (commented):
A few basic questions.
* Does this need to be documented anywhere? Is there a help command we already have that shows available switches?
* What if I pass in a bad url? What kind of feedback does the user get?

Changes look OK overall; I was wondering about some of those other aspects.

arizzitano reviewed (commented):
I'd recommend adding documentation for the `single_url` option in the options table in the README. pa11ycrawler itself doesn't have an explicit help command (it defaults to scrapy's), but we should also make sure `single_url` is adequately documented in edx-platform over on https://github.com/edx/edx-platform/pull/14485.

I'd encourage some documentation in https://openedx.atlassian.net/wiki/display/A11Y/pa11ycrawler+Overview, as well, after both PRs have landed.

michaelyoungstrom commented:
**Update:**

Ben, regarding your feedback I have made the following changes:
- Added documentation inside the README to provide some information about this feature.
- Included a new `handle_error` method that will provide a little more information when a single_url request goes wrong (such as a bad url).

Outside of the suggested feedback, I also cleaned up a new issue with fetching credentials. Previously, credentials were obtained via auto_auth any time there was a `single_url` argument passed into the spider. However there were a few problems with that. 
- If credentials are passed into the spider, there is no need to go through auto_auth
- If the `single_url` did not use the default port `8000`, there was be an error unless they passed the valid port number into the spider
- If a url was outside of localhost (say, google.com) there obviously is no auto_auth support

The solution to this can be reviewed inside of the `start_requests` method.

benpatterson commented on README.md:
```diff
@@ -70,6 +71,10 @@ so if you want to clear it out between runs, that's your responsibility.
 `rm -rf data`.
 
+The `single_url` option is available to allow the spider to only crawl one web page. 
+The result is treated through the pipeline, but the spider will not continue crawling
```
nit: I would change "treated" to "evaluated"

benpatterson commented on pa11ycrawler/spiders/edx.py:
```diff
@@ -5,6 +5,10 @@
 import re
 import json
 from datetime import datetime
+try:
```
nit: I'd add a comment on why this is here (i'm guessing it's py2 vs py3?)

benpatterson commented on pa11ycrawler/spiders/edx.py:
```diff
@@ -144,6 +159,15 @@ def start_requests(self):
+                if port != self.port:
+                    self.port = port
+            else:
+                # No need for credentials
```
I wonder if a log message would be helpful here.

benpatterson commented on pa11ycrawler/spiders/edx.py:
```diff
@@ -215,13 +239,17 @@ def after_initial_login(self, response):
         """
-        if "We couldn't sign you in." in response.text:
+        if LOGIN_FAILURE_MSG in response.text:
             self.logger.error("Credentials failed!")
```
I wonder if you could make the error more helpful (e.g., include suggestions for what to do to avoid the problem, like either include your credentials or log in first)

benpatterson commented on tests/test_spider.py:
```diff
@@ -200,6 +200,31 @@ def test_load_pa11y_rules_file(tmpdir):
     assert result == expected_result
 
 
+def test_single_url():
```
Could you check your coverage numbers? Have you covered the code you introduced? (I don't know...just asking. One test seems like it's too little but any scenario I can think of would likely be covered by other existing tests.)

benpatterson commented:
@michaelyoungstrom before merging, you'll want to squash your commits and rebase. When merging, you'll also want to re-title the request so that "WIP" isn't going to appear in the commit logs. (There will be a merge commit when you merge.) Let us know if you want a hand with any of these; it could be 2 minutes of showing you how to do it vs grok-time on your own ;)
